02-microarray/pathway-analysis_microarray_04_gsva.Rmd

```diff
@@ -0,0 +1,631 @@
+head(mapped_df)
+```
+
+We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as rownames.
+In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.
+
+Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+Let's find out which Entrez IDs have been duplicated.
+
+```{r}
+dup_entrez_ids <- mapped_df %>%
+  dplyr::filter(duplicated(entrez_id)) %>%
+  dplyr::pull(entrez_id)
+
+dup_entrez_ids
+```
+
+Now let's take a look at the rows associated with the duplicated Entrez IDs.
+
+```{r}
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids)
+```
+
+We can see that the associated values vary for each row.
+
+As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the larger mean expression value per sample.
+
+We are removing values for three genes here, so it is unlikely to have a considerable impact on our results.
+
+**REVIEW**
+
+```{r}
+# First let's create a column with the mean expression value for each row -- we
+# will want to calculate the mean using the numeric expression values only so
+# here we use negative indexing to account for the two gene identifier columns
+# from the previous mapping step that we do not want included in the calculation
+# of the mean
+mapped_df$mean <- apply(mapped_df[, -c(1:2)], 1, mean)
+
+# Let's look at the rows associated with the duplicated Entrez IDs now that we
+# have a mean expression value
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
+  dplyr::select(Ensembl, entrez_id, mean, dplyr::everything())
+```
+
+Now we can filter out the duplicate gene identifiers using the mean expression value.
+
+```{r}
+filtered_mapped_df <- mapped_df %>%
+  # Sort so that the highest absolute mean expression values are at the top
+  dplyr::arrange(dplyr::desc(abs(mean))) %>%
+  # Filter out the duplicated rows using `dplyr::distinct()`-- this will keep
+  # the first row with the duplicated value thus keeping the row with the
+  # highest absolute value of the mean expression
+  dplyr::distinct(entrez_id, .keep_all = TRUE)
+```
+
+Let's check to see that we removed the duplicate Entrez IDs and kept the rows with the higher absolute value of the mean expression.
+
+```{r}
+filtered_mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
+  dplyr::select(Ensembl, entrez_id, mean, dplyr::everything())
+```
+
+Looks like we were able to successfully get rid of the duplicate gene identifiers and keep the observations with the higher absolute mean expression value!
+
+As mentioned earlier, we need a matrix for GSVA.
+Let's now convert our data frame into a matrix and prepare our object for GSVA.
+
+```{r}
+matrix <- filtered_mapped_df %>%
```

Comment:
This name has to change because it will conflict with `matrix()`. It's also not very descriptive!